### PR TITLE
infra: Move unit tests to a subdirectory to allow space for integration tests

### PR DESCRIPTION
**Pull request by malexw**
As part of a larger series of testing improvements, this change moves all the current python tests from `backend/tests` to the subdirectory `backend/tests/unit`. This gives us room to add a directory for integration tests with their own config.

**AI Description**

This PR introduces changes to the Makefile and updates several test files.

## Makefile
- Adds new targets: `watch`, `up`, `down`, `run-unit-tests`, `run-community-tests`, `run-integration-tests`, `attach`, `exec-backend`, `exec-db`, `migration`, `migrate`, `downgrade`, `reset-db`, `setup`, `setup-use-community`, `win-setup`, `lint`, `first-run`, `win-first-run`, `generate-client-web`, and `install-web`.
- Updates the `run-tests` target to include `run-unit-tests`.
- Removes the `run-tests` target's dependency on `build`.

## Test Files
- Moves several test files from `src/backend/tests/` to `src/backend/tests/unit/`: configuration.yaml, conftest.py, test_deployments.py, test_agent.py, test_agent_tool_metadata.py, test_blacklist.py, test_citation.py, test_conversation.py, test_document.py, test_file.py, test_message.py, test_organization.py, test_snapshot.py, test_tool_call.py, and test_user.py.
- Updates import statements in the moved test files to use the new path.
- Removes the `src/backend/tests/factories/` and `src/backend/tests/model_deployments/mock_deployments/` directories and their contents.
- Adds a new file, `src/backend/tests/unit/factories/__init__.py`, containing the same contents as the removed `src/backend/tests/factories/__init__.py`.

### CODE REVIEW

File: .github/workflows/backend_integration_tests.yml [hunk 1]
```diff
@@ -20,12 +20,12 @@ jobs:
         uses: actions/checkout@v3
       - name: Setup Docker
         run: docker compose up test_db -d
-      - name: Run tests
+      - name: Run unit tests
```
Comment (BeatrixCohere) [inline .github/workflows/backend_integration_tests.yml:23]:
```suggestion
      - name: Run integration tests
```

File: .github/workflows/backend_unit_tests.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Checkout repo
+        uses: actions/checkout@v3
+      - name: Setup Docker
+        run: docker compose up test_db -d
+      - name: Run unit tests
+        if: github.actor != 'dependabot[bot]'
+        run: |
+          mkdir -p coverage
+          docker compose build --build-arg INSTALL_COMMUNITY_DEPS=true backend
+          docker compose run --name backend-container -v $(pwd)/coverage:/app/coverage -e COHERE_API_KEY=$COHERE_API_KEY backend poetry run pytest --cov=src/backend --cov-report=xml:/app/coverage/coverage.xml src/backend/tests/unit
```
Comment (BeatrixCohere) [inline .github/workflows/backend_unit_tests.yml:28]: Does it need the API key?

File: .github/workflows/backend_unit_tests.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Upload coverage reports to Codecov
+        uses: codecov/codecov-action@v4.0.1
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (BeatrixCohere) [inline .github/workflows/backend_unit_tests.yml:32]: Can we start without any secrets?

File: .github/workflows/backend_unit_tests.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+      - name: Setup Docker
+        run: docker compose up test_db -d
+      - name: Run unit tests
+        if: github.actor != 'dependabot[bot]'
+        run: |
+          mkdir -p coverage
+          docker compose build --build-arg INSTALL_COMMUNITY_DEPS=true backend
```
Comment (BeatrixCohere) [inline .github/workflows/backend_unit_tests.yml:27]: Do unit tests still need an service to run? Could we just not run them independently

File: .github/workflows/backend_integration_tests.yml (hunk 1 above)
Comment (malexw) [reply .github/workflows/backend_integration_tests.yml]: Good eye!

File: .github/workflows/backend_unit_tests.yml (hunk 2 above)
Comment (malexw) [reply .github/workflows/backend_unit_tests.yml]: It turns out we don't

File: .github/workflows/backend_unit_tests.yml (hunk 3 above)
Comment (malexw) [reply .github/workflows/backend_unit_tests.yml:32]: As far as I can tell, in order to have Codecov monitor our coverage, we do need to provide the API key

File: .github/workflows/backend_unit_tests.yml (hunk 4 above)
Comment (malexw) [reply .github/workflows/backend_unit_tests.yml:27]: I think you're asking if we could run the tests without needing to build the docker container, right?

That is how I run single tests or single test files locally, so it's certainly achievable. I see that our tooling also builds a container and runs the tests there when we use the `make run-unit-tests` command, so I'm guessing whoever set this workflow up decided to do things here the same way either for consistency or for ease of implementation.

**Review by BeatrixCohere — APPROVED** [review]

**Review by tomtobac — APPROVED** [review]

**Pull request merged**